src/main/java/htsjdk/samtools/BAMShardIndexer.java

```diff
@@ -0,0 +1,190 @@
+                  final Integer binNumber = rec.getIndexingBin();
+                  return (binNumber == null ? rec.computeIndexingBin() : binNumber);
+
+              }
+
+              @Override
+              public Chunk getChunk() {
+                  final SAMFileSource source = rec.getFileSource();
+                  if (source == null) {
+                      throw new SAMException("No source (virtual file offsets); needed for indexing on BAM Record " + rec);
+                  }
+                  return ((BAMFileSpan) source.getFilePointer()).getSingleChunk();
+              }
+          });
+
+      }
+
+      /**
+       * Creates the BAMIndexContent for this reference.
+       * Requires all alignments of the reference have already been processed.
+       *
+       * @return Null if there are no features for this reference.
+       */
+      public BAMIndexContent processReference(final int reference) {
+
+          if (reference != currentReference) {
+              throw new SAMException("Unexpected reference " + reference + " when constructing index for " + currentReference);
+          }
+
+          final BinningIndexContent indexContent = binningIndexBuilder.generateIndexContent();
+          if (indexContent == null) return null;
+          return new BAMIndexContent(indexContent.getReferenceSequence(), indexContent.getBins(),
+                  indexStats, indexContent.getLinearIndex());
+
+      }
+
+      /**
+       * @return the count of records with no coordinate positions
+       */
+      public long getNoCoordinateRecordCount() {
+          return indexStats.getNoCoordinateRecordCount();
+      }
+
+      /**
+       * reinitialize all data structures when the reference changes
+       */
+      void startNewReference(int reference) {
+          currentReference  = reference;
+          // I'm not crazy about recycling this object, but that is the way it was originally written and
+          // it helps keep track of no-coordinate read count (which shouldn't be stored in this class anyway).
+          indexStats.newReference();
+          binningIndexBuilder = new BinningIndexBuilder(currentReference,
+                  sequenceDictionary.getSequence(currentReference).getSequenceLength());
+      }
+  }
+
+  /**
+   * Generates a BAM index file from an input BAM file
+   *
+   * @param reader SAMFileReader for input BAM file
+   * @param output File for output index file
+   */
+  public static void createIndex(SamReader reader, File output) {
+      createIndex(reader, output, null);
+  }
+
+  /**
+   * Generates a BAM index file from an input BAM file
+   *
+   * @param reader SAMFileReader for input BAM file
+   * @param output File for output index file
+   */
+  public static void createIndex(SamReader reader, File output, Log log) {
+
+      BAMIndexer indexer = new BAMIndexer(output, reader.getFileHeader());
+
+      int totalRecords = 0;
+
+      // create and write the content
```

Comment:
s/create and write the content/Populates the index from the records in a BAM file.